src/Capability/Reflection.hs

```diff
@@ -0,0 +1,170 @@
+import Data.Reflection
+
+-- | @interpret_ \@tag dict action@
+--
+-- Execute @action@ using the ad-hoc interpretation of a capability @c@ under
+-- @tag@ defined by @dict@, where @dict@ is a value of type @'Reified' tag c@,
+-- i.e. a record providing the implementation of the methods of capability @c@.
+--
+-- For example, the following provides an ad-hoc interpretation for the
+-- 'Capability.Source.HasSource' capability.
+--
+-- >>> :{
+--   interpret_ @"my-source"
+--     ReifiedSource { _await = pure "capabilities" }
+--     (replicateM 3 (await @"my-source"))
+--   :}
+-- ["capabilities", "capabilities", "capabilities"]
+interpret_ ::
+  forall tag c m a.
+  ( Monad m,
+    forall s. Reifies s (Reified tag c m) => c (Reflected s c m)
+  ) =>
+  Reified tag c m ->
+  (forall m'. c m' => m' a) ->
+  m a
+interpret_ = interpret @tag @'[] @c
+{-# INLINE interpret_ #-}
+
+-- | @interpret \@tag \@ambient dict action@
+--
+-- Like 'interpret_' but allows to forward the ambient capabilities @ambient@
+-- into the context of @action@ as well.
+--
+-- For example, the following provides an ad-hoc interpretation for the
+-- 'Capability.Source.HasSource' capability, while using an ambient
+-- 'Capability.Sink.HasSink' capability.
+--
+-- >>> :{
+--   interpret @"my-source" @'[HasSink "my-sink" String]
+--     ReifiedSource { _await = pure "capabilities" }
+--     (replicateM_ 3 (await @"my-source" >>= yield @"my-sink"))
+--   :}
+interpret ::
+  forall tag (cs :: [Capability]) c m a.
+  ( Monad m,
+    All cs m,
+    forall s. Reifies s (Reified tag c m) => c (Reflected s c m)
+  ) =>
+  Reified tag c m ->
+  (forall m'. All (c ': cs) m' => m' a) ->
+  m a
+interpret dict action =
+  reify dict $ \(_ :: Proxy s) ->
+    derive @(Reflected s c) @'[c] @cs action
+{-# INLINE interpret #-}
+
+-- | @Reified tag capability m@
+--
+-- Defines the dictionary type for the methods of @capability@ under @tag@ in
+-- the monad @m@. Refer to 'interpret_' for an example use-case.
+--
+-- For example, the 'Capability.Sink.HasSink' capability has the method
+-- @'Capability.Sink.yield' :: a -> m ()@. The corresponding dictionary type is
+-- defined as follows.
+--
+-- >>> :{
+--   data instance Reified tag (HasSink tag a) m =
+--     ReifiedSink { _yield :: forall a. a -> m () }
+--   :}
+--
+-- Superclass dictionaries are represented as nested records. For example, the
+-- 'Capability.State.HasState' capability has the superclasses
+-- 'Capability.Source.HasSource' and 'Capability.Sink.HasSink' and the method
+-- @'Capability.State.state' :: (s -> (a, s)) -> m a@. The corresponding
+-- dictionary type is defined as follows.
+--
+-- >>> :{
+--   data instance Reified tag (HasState tag s) m =
+--     ReifiedState
+--       { _stateSource :: Reified tag (HasSource tag s) m,
+--         _stateSink :: Reified tag (HasSink tag s) m,
+--         _state :: forall a. (s -> (a, s)) -> m a
+--       }
+--   :}
+data family Reified (tag :: k) (c :: Capability) (m :: * -> *)
```

Comment:
I kind of feel we should generalize this data family and add it to its own package. (or to reflection possibly), with the intent that we could generate these with Template Haskell, probably:

```haskell
data family Reified (c :: Constraint)
```

(and we would transform `Reified tag c m` into `Reified (c m)`.

The tag disappears, in this transformation. This is fishy, do we actually need the tag in the current presentation?

Reply:
`Reified` was a convenient location to attach the `tag` for `interpret_` and `interpret` to ship it into the `action`. The suggested transformation is possible, then we just need to find another way to transport the `tag` into the action of `interpret`.

What I mean is the following. Take [this example](https://github.com/tweag/capability/commit/4fe654033a53147630308c71b3f3a5e035363c7b):
```
example :: [()]
example =
  interpret_ @"sink"
    ReifiedSink { _yield = (:[]) }
    (do yield @"sink" (); yield @"sink" ())
```
And assume 
```
data family Reified (c :: Constraint)
interpret_ :: forall tag c m a.
  (Monad m, forall s. Reifies s (Reified (c m)) => c (Reflected s c m)) =>
  Reified (c m) -> (forall m'. c m' => m' a) -> m a
```
Then we will get the following type error on `example`:
```
    • Could not deduce (HasSink "sink" () m')
        arising from a use of ‘yield’
      from the context: HasSink tag0 () m'
        bound by a type expected by the context:
                   forall (m' :: * -> *). HasSink tag0 () m' => m' ()
```
I.e. it hasn't inferred `tag0 ~ "sink"` because there is no such constraint between `tag` and `c`.

A user would have to provide a type annotation to the `ReifiedSink` constructor to set the `tag`, [for example](https://github.com/tweag/capability/commit/a24d9453c3b68fb5d8f4849a0c6b0fa7e473ae09)
```
  interpret_
    (ReifiedSink { _yield = (:[]) } :: Reified (HasSink "sink" () []))
    (do yield @"sink" (); yield @"sink" ())
```
That's pretty clumsy to write. Is there a way to use `TypeApplications` on record constructors? I tried the following, which don't work
```
ReifiedSink @_ @"sink" { _yield = ... }
ReifiedSink { _yield = ... } @_ @"sink"
```

I've found two other ways that allow the same API to `interpret_`.

1. Be explicit about the `tag` in the capability, i.e. use the following type for [`interpret_`](https://github.com/tweag/capability/commit/cee96406f23f0d0045f7999fec069af2c0c1d46c):
    ```
    interpret_ :: forall tag c m v a.
      (Monad m, forall s. Reifies s (Reified (c tag v m)) => c tag v (Reflected s (c tag v) m)) =>
      Reified (c tag v m) -> (forall m'. c tag v m' => m' a) -> m a
    ```
    I'm not too fond of this approach because it hard codes that capabilities should have the kind `tag -> value -> (* -> *) -> Constraint`. This happens to be the case so far, because all capabilities have some sort of value type parameter, e.g. the `r` in `HasReader tag r` or the `w` in `HasWriter tag w`. But, this seems coincidental. Why shouldn't there be a capability with none, or two, or more value type parameters?
    Maybe the shape of capabilities should have been `value -> tag -> (* -> *) -> Constraint`, e.g. `HasReader r tag`, then we could use `c tag m` in `interpret_` without making assumptions about the number of value type parameters.
2. Define an additional constraint, say [`TagOf tag c`](https://github.com/tweag/capability/commit/46c2148aba56a1d73050aec0a71eafc30a1bdf77), and use it as follows
    ```
    interpret_ :: forall tag c m a.
      (TagOf tag c, Monad m, forall s. Reifies s (Reified (c m)) => c (Reflected s c m)) =>
      Reified (c m) -> (forall m'. (TagOf tag c, c m') => m' a) -> m a
    ```
    This enforces that `c` has the `tag` without hard-coding the number of value type parameters.
    This requires some boilerplate when defining a new capability though, e.g. instances of the form
    ```
    instance tag ~ tag' => TagOf tag (HasReader tag' r)
    ```
    (The [example](https://github.com/tweag/capability/commit/46c2148aba56a1d73050aec0a71eafc30a1bdf77) takes a short cut with `instance tag ~ tag' => TagOf tag (c tag' v)`, but that seems to conflict with instances for capabilities with different number of value type parameters.)

Out of all those, attaching `tag` to `Reified` seemed like the least bad option.